### ISSUE TITLE
FP daemon gRPC is blocked before fast sync finish (low-pri)

### ISSUE THREAD

**Issue by bap2pecs** [babylonlabs-io/finality-provider#29]
When we were testing our OP integration, we started fpd but the 12581 RPC port never gets ready:

```
Whoops. There was an error while executing your fpd CLI 'rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:12581: connect: connection refused"'%     
```

after turning on `debug` level log, we found the root cause is the daemon won't start gRPC server until fast sync is complete

code:

in `finality-provider/cmd/fpd/daemon/start.go`:
```
func runStartCmd(ctx client.Context, cmd *cobra.Command, args []string) error {
        ...

        // this triggers fast sync
	if err := startApp(fpApp, fpStr, passphrase); err != nil {
		return fmt.Errorf("failed to start app: %w", err)
	}

	...
	
	// this starts gRPC server
	return fpServer.RunUntilShutdown()
```

call stack:
`startApp() -> fpApp.StartHandlingAll() -> app.fpManager.StartAll() -> fpm.StartFinalityProvider() -> fpm.addFinalityProviderInstance() -> fpIns.Start() -> fp.bootstrap() -> fp.tryFastSync() -> fp.FastSync()`

in FastSync(), it has this blocking for-loop:

```
func (fp *FinalityProviderInstance) FastSync(startHeight, endHeight uint64) (*FastSyncResult, error) {
	...
	for startHeight <= endHeight {
		blocks, err := fp.consumerCon.QueryBlocks(startHeight, endHeight, fp.cfg.FastSyncLimit)
		...
	}
	...
}
```

**Comment by bap2pecs** [comment]
cc @SebastianElvis @gitferry

**Comment by SebastianElvis** [comment]
I remember this is expected as before that information in FP is not updated?

**Comment by bap2pecs** [comment]
> I remember this is expected as before that information in FP is not updated?

right but this means even `fpd list-finality-providers` will fail before fast sync is fully finished.

**Comment by SebastianElvis** [comment]
Yep it does, along with some other commands. Fully sync typically does not take long so maybe we are good?

**Comment by bap2pecs** [comment]
> Yep it does, along with some other commands. Fully sync typically does not take long so maybe we are good?

sure but one possible scenario is the FP gets offline for a long time. in that case, fast sync can be very slow?

**Comment by gitferry** [comment]
@bap2pecs sorry for being late here. So in normal case, fast sync would not be an issue because it sync from the last finalized block. Nevertheless, to solve this issue, can we move the start of the finality app into `server.RunUntilShutdown` ahead of grpc server being started?

**Comment by bap2pecs** [comment]
> to solve this issue, can we move the start of the finality app into `server.RunUntilShutdown` ahead of grpc server being started?

but the problem is starting the finality app will trigger fast sync. so I think we need to first start the grpc then starting the finality app to avoid this issue. wdyt?

**Comment by gitferry** [comment]
I was trying to say the app should start after the grpc. But after a second thought, we should not start grpc server before the app is fully initialized because if we do we would expose out-of-sync data

**Comment by bap2pecs** [comment]
> we should not start grpc server before the app is fully initialized because if we do we would expose out-of-sync data

makes sense. but I think a potential improvement is to only allow certain requests (e.g. list FP) before it's done syncing or throw error for certain requests before it's done syncing

**Comment by gitferry** [comment]
list FP should show the info of FP (e.g., last voting height) which should be queried after fast sync. So for a better UX, we can start grpc server before app is initialized but we should return error for requests like `the finality provider is still syncing` before the syncing is done. Low priority I would say

**Comment by bap2pecs** [comment]
agree it's a low-pri. I will leave this issue up for grab. probably a good ramp-up task for future team members

**Comment by bap2pecs** [comment]
fast sync is removed
